Provide new API to programmatically create REST Client reactive

Sgitario opened the pull request:
The new API includes these additional options:

- verifyHost
- trustStore by truststore and password (it was requested in #31891)
- proxyUser
- proxyPassword
- nonProxyHosts

Note that for the new API, I didn't add any breaking change, so users can keep using the Microprofile one.

Fix https://github.com/quarkusio/quarkus/issues/32856

geoand commented:
We'll need a resolution on https://github.com/quarkusio/quarkus/issues/32856#issuecomment-1520077546 before moving this forward

geoand commented on extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/QuarkusRestClientBuilder.java:
```diff
@@ -0,0 +1,236 @@
+     * @return the current builder with the connect timeout set.
+     * @throws IllegalArgumentException if the value of timeout is negative.
+     */
+    QuarkusRestClientBuilder readTimeout(long timeout, TimeUnit unit);
```
We should probably also have tests where these properties override the configured values in `application.properties`

Sgitario replied:
Removed

Sgitario replied:
This is covered by RestClientCDIDelegateBuilderTest

Sgitario commented on extensions/resteasy-reactive/rest-client-reactive/runtime/src/main/java/io/quarkus/rest/client/reactive/QuarkusRestClientBuilder.java:
```diff
@@ -0,0 +1,236 @@
+
+    static QuarkusRestClientBuilder newBuilder() {
+        RestClientBuilderImpl proxy = new RestClientBuilderImpl();
+        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+            for (RestClientBuilderListener listener : ServiceLoader.load(RestClientBuilderListener.class)) {
+                listener.onNewBuilder(proxy);
+            }
+            return null;
+        });
```
Removed

geoand commented on docs/src/main/asciidoc/rest-client-reactive.adoc:
```diff
@@ -394,7 +394,7 @@ public class ExtensionsResource {
 
     public ExtensionsResource() {
-        extensionsService = RestClientBuilder.newBuilder()
+        extensionsService = QuarkusRestClientBuilder.newBuilder()
```
Maybe we should add a call out explaining this?

geoand replied:
No no, I mean that in the docs we should explain the use of this new builder vs the Microprofile one.

Now users just see the class in snippets of code so might not realize what is going on.

Furthermore, experienced users will likely never see these snippets.

If instead we have a `TIP` or something, it will be a lot more visible

Sgitario replied:
+1 will update it

Sgitario replied:
PR updated.

Sgitario committed:
Provide new API to programmatically create REST Client reactive

The new API includes these additional options:

- verifyHost
- trustStore by truststore and password (it was requested in #31891)
- proxyUser
- proxyPassword
- nonProxyHosts

Fix https://github.com/quarkusio/quarkus/issues/32856

geoand reviewed (approved):
👌
